libdrgn/iterators.c

```diff
@@ -0,0 +1,461 @@
+	} else if (err->code == DRGN_ERROR_LOOKUP) {
+		err = pid_iter_init_old_kernel(prog, ns, ret);
+	}
+
+out:
+	drgn_object_deinit(&idr);
+	return err;
+}
+
+struct drgn_error *pid_iter_next(struct pid_iter *iter, pid_iter_entry_t **ret)
+{
+	struct drgn_error *err;
+	if (iter->has_idr) {
+		idr_iter_entry_t *entry;
+		err = idr_iter_next(&iter->iter, &entry);
+		if (err)
+			return err;
+		if (!entry) {
+			*ret = NULL;
+			return NULL;
+		}
+		err = drgn_object_cast(&entry->node, iter->pid_type,
+				       &entry->node);
+		if (!err)
+			*ret = &entry->node;
+		return err;
+	} else {
+		struct drgn_object upid;
+		struct drgn_object upid_ns;
+		bool done = false;
+		drgn_object_init(&upid, drgn_object_program(&iter->ns));
+		drgn_object_init(&upid_ns, drgn_object_program(&iter->ns));
+
+		while (!done &&
+		       (iter->index == -1 || iter->index < iter->hash_length)) {
+			bool is_not_null;
+			err = drgn_object_bool(&iter->pos, &is_not_null);
+			if (err)
+				return err;
+			if (!is_not_null) {
+				err = drgn_object_subscript(&iter->pos,
+							    &iter->pid_hash,
+							    ++iter->index);
+				if (err)
+					return err;
+				err = drgn_object_member(&iter->pos, &iter->pos,
+							 "first");
+				if (err)
+					return err;
+				continue;
+			}
+			err = drgn_object_container_of(&upid, &iter->pos,
+						       iter->upid_type,
+						       "pid_chain");
+			if (err)
+				return err;
+			err = drgn_object_member_dereference(&upid_ns, &upid,
+							     "ns");
+			if (err)
+				return err;
+			int ns_cmp_result;
+			err = drgn_object_cmp(&upid_ns, &iter->ns,
+					      &ns_cmp_result);
+			if (err)
+				return err;
+			if (ns_cmp_result == 0) {
+				err = drgn_object_container_of(
+					&iter->entry, &upid, iter->pid_type,
+					iter->member_specifier);
+				if (err)
+					return err;
+				done = true;
+				*ret = &iter->entry;
+			}
+			err = drgn_object_member_dereference(
+				&iter->pos, &iter->pos, "next");
+			if (err)
+				return err;
+		}
+		if (!done)
+			*ret = NULL;
+		return NULL;
+	}
+}
+
+void pid_iter_deinit(struct pid_iter *iter)
+{
+	if (iter->has_idr) {
+		idr_iter_deinit(&iter->iter);
+	} else {
+		drgn_object_deinit(&iter->pid_hash);
+		drgn_object_deinit(&iter->pos);
+		drgn_object_deinit(&iter->ns);
+		drgn_object_deinit(&iter->entry);
+	}
+}
+
+struct drgn_error *task_iter_init(struct drgn_program *prog,
+				  const struct drgn_object *ns,
+				  struct task_iter *ret)
+{
+	struct drgn_error *err = pid_iter_init(prog, ns, &ret->iter);
+	if (err)
+		return err;
+	struct drgn_object PIDTYPE_PID;
+	drgn_object_init(&PIDTYPE_PID, prog);
+	err = drgn_program_find_object(prog, "PIDTYPE_PID", NULL,
+				       DRGN_FIND_OBJECT_CONSTANT, &PIDTYPE_PID);
+	if (err)
+		return err;
+	err = drgn_object_read_unsigned(&PIDTYPE_PID, &ret->PIDTYPE_PID);
+	return err;
+}
+
+struct drgn_error *task_iter_next(struct task_iter *iter,
+				  task_iter_entry_t **ret)
+{
+	struct drgn_error *err;
+	bool value_is_not_null;
```

Comment:
Personally, I'd prefer the name `truthy` for this rather than the double negative of !is_not_null.

Reply:
Would introducing `drgn_object_is_null` to `drgn/libdrgn/object.c` be a satisfactory alternative? This way the variable name could simply be `bool value_is_null`, avoiding the double negative while still making it explicit that we're guarding against a `NULL` pointer (as opposed to the more generic test for "truthiness".

Reply:
I'd rather not add another operator for what is essentially a stylistic choice between `if (ptr == NULL)` and `if (!ptr)`.